contracts/protocol/clients/priceDiscovery/BosonPriceDiscovery.sol

```diff
@@ -0,0 +1,275 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.22;
+
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { ERC165 } from "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+import { IWrappedNative } from "../../../interfaces/IWrappedNative.sol";
+import { IBosonVoucher } from "../../../interfaces/clients/IBosonVoucher.sol";
+import { IBosonPriceDiscovery } from "../../../interfaces/clients/IBosonPriceDiscovery.sol";
+import { IERC721Receiver } from "../../../interfaces/IERC721Receiver.sol";
+import { FundsLib } from "../../libs/FundsLib.sol";
+import { Address } from "@openzeppelin/contracts/utils/Address.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { BosonTypes } from "../../../domain/BosonTypes.sol";
+import { BosonErrors } from "../../../domain/BosonErrors.sol";
+
+/**
+ * @title BosonPriceDiscovery
+ *
+ * @dev Boson Price Discovery is a external contract that is used to determine the price of an exchange.
+ */
+contract BosonPriceDiscovery is ERC165, IBosonPriceDiscovery, BosonErrors {
+    using Address for address;
+    using SafeERC20 for IERC20;
+
+    IWrappedNative internal immutable wNative;
+
+    address private incomingTokenAddress;
+
+    address private immutable bosonProtocolAddress;
+
+    /**
+     * @notice
+     * For offers with native exchange token, it is expected the the price discovery contracts will
+     * operate with wrapped native token. Set the address of the wrapped native token in the constructor.
```

Comment:
```suggestion
     * operate with wrapped native token. Sets the address of the wrapped native token in the constructor.
```

Reply:
I kept that as it was, since it's meant as an instruction for the deployer.